Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -0,0 +1,124 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Computability.Language
+
+/-!
+# Context-Free Grammar
+
+This file contains the definition of a Context-Free Grammar, which is a grammar that has a single
+nonterminal symbol on the left-hand side of each rule.
+-/
+
+/-- Rule that rewrites a single nonterminal to any list of symbols. -/
+structure ContextFreeRule (T : Type*) (N : Type*) where
+  /-- Input nonterminal a.k.a. left-hand side -/
+  input : N
+  /-- Output string a.k.a. right-hand side -/
+  output : List (Symbol T N)
+
+/-- Context-free grammar that generates words over the alphabet `T` (a type of terminals). -/
+structure ContextFreeGrammar (T : Type*) where
+  /-- Type of nonterminals -/
+  NT : Type*
+  /-- Initial nonterminal -/
+  initial : NT
+  /-- Rewrite rules -/
+  rules : List (ContextFreeRule T NT)
+
+variable {T : Type*}
+
+/-- One application of single context-free rule. -/
+inductive ContextFreeRule.RewritesTo {N : Type*} (r : ContextFreeRule T N) :
+    List (Symbol T N) → List (Symbol T N) → Prop
+  /-- The replacement is at the start of the remaining string. -/
+  | head (s : List (Symbol T N)) :
+      r.RewritesTo (Symbol.nonterminal r.input :: s) (r.output ++ s)
+  /-- There is a replacement later in the string. -/
+  | cons (x : Symbol T N) {s₁ s₂ : List (Symbol T N)} (hrs : RewritesTo r s₁ s₂) :
+      r.RewritesTo (x :: s₁) (x :: s₂)
+
+lemma ContextFreeRule.RewritesTo.exists_parts {N : Type _} {r : ContextFreeRule T N}
```

Comment:
```suggestion
lemma ContextFreeRule.RewritesTo.exists_parts {N : Type*} {r : ContextFreeRule T N}
```
and throughout

Reply:
application type mismatch
  RewritesTo ?m.1518 ?m.1539 v
argument
  v
has type
  List (Symbol T N) : Type (max u_2 u_1)
but is expected to have type
  List (Symbol ?m.1486 ?m.1487) : Type u_2

Reply:
I wasn't able to reproduce this